File: heap-web/src/main/java/com/github/ol_loginov/heaplibweb/services/proxies/ObjectFieldValueProxy.java
package com.github.ol_loginov.heaplibweb.services.proxies;

import com.github.ol_loginov.heaplibweb.repository.heap.FieldValueEntity;
import com.github.ol_loginov.heaplibweb.repository.heap.HeapScope;
import com.github.ol_loginov.heaplibweb.repository.heap.InstanceEntity;
import org.netbeans.lib.profiler.heap.Instance;
import org.netbeans.lib.profiler.heap.ObjectFieldValue;

public class ObjectFieldValueProxy extends FieldValueProxy implements ObjectFieldValue {
	public ObjectFieldValueProxy(int heapId, FieldValueEntity entity, HeapScope scope) {
		super(heapId, entity, scope);
	}

	@Override
	public Instance getInstance() {
		return scope.getInstances().findById(new InstanceEntity.PK(heapId, entity.getDefiningInstanceId()))
			.map(e -> new InstanceProxy(e, scope))
			.orElse(null);
	}

	@Override
	public long getInstanceId() {
		return entity.getValueInstanceId();
	}
}

File: heap-web/src/main/java/com/github/ol_loginov/heaplibweb/services/proxies/ObjectFieldValueProxy.kt
package com.github.ol_loginov.heaplibweb.services.proxies

import com.github.ol_loginov.heaplibweb.repository.heap.FieldValueEntity
import com.github.ol_loginov.heaplibweb.repository.heap.HeapRepositories
import org.netbeans.lib.profiler.heap.Instance
import org.netbeans.lib.profiler.heap.ObjectFieldValue

class ObjectFieldValueProxy(
    entity: FieldValueEntity,
    heapRepositories: HeapRepositories
) : FieldValueProxy(entity, heapRepositories), ObjectFieldValue {
    override fun getInstance(): Instance? = heapRepositories
        .instances.findById(entity.valueInstanceId)
        ?.let { InstanceProxy(it, heapRepositories) }

    override fun getInstanceId(): Long = entity.valueInstanceId
}